File: SWEA/D5/[ 01247 ] 최적경로.java
import java.util.Scanner;

public class Solution {
    static int MIN;
    static int N;
    static int[][] arr;
    static boolean[] visit;
    static int [][] inner;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        sc.nextLine();
        for(int test = 0; test<= T; test++){
            N = sc.nextInt();
            sc.nextLine();
            String[] arr_ = sc.nextLine().split(" ");
            arr = new int[N+2][2];
            for(int n = 0; n<N+2; n++)
                arr[n] = new int[]{Integer.parseInt(arr_[2*n]), Integer.parseInt(arr_[2*n+1])};
            visit = new boolean[N+2];
            inner = new int[N+2][N+2];
            MIN = Integer.MAX_VALUE;

            for(int i = 0; i<N+2; i++){
                for(int j = i+1; j<N+2; j++){
                    inner[i][j] = Math.abs(arr[i][0] - arr[j][0]) + Math.abs(arr[i][1] - arr[j][1]);
                    inner[j][i] = Math.abs(arr[i][0] - arr[j][0]) + Math.abs(arr[i][1] - arr[j][1]);
                }
            }
            dfs(0,0,0);
            System.out.println("#"+test+" " + MIN);
        }
    }

    static void dfs(int now, int count, int total){
        if (count == N) {
            total += inner[now][1];
            MIN = Math.min(MIN, total);
            return;
        }

        if(MIN < total)
            return;

        for(int i = 2; i<N+2; i++) {
            if(visit[i])
                continue;
            visit[i] =true;
            dfs(i, count + 1, total + inner[now][i]);
            visit[i] = false;
        }
    }
}

File: SWEA/D5/[ 01247 ] 최적경로.py
# https://swexpertacademy.com/main/talk/solvingClub/problemView.do?solveclubId=AWv0A3daT1QDFAW4&contestProbId=AV15OZ4qAPICFAYD&probBoxId=AWyzMqVathYDFAVP&type=PROBLEM&problemBoxTitle=0821_day11_ws&problemBoxCnt=1

def dfs(now, count, total):
    global MIN
    if count == N:
        total += inner[now][1]
        MIN = min(MIN, total)
        return

    if MIN < total:
        return

    for i in range(2,N+2):
        if visit[i]:
            continue
        visit[i] = True
        dfs(i, count+1, total + inner[now][i])
        visit[i] = False


T = int(input())
for test in range(1,T+1):
    N = int(input())
    arr_ = list(map(int, input().split()))
    arr = []
    for n in range(0,N+2):
        arr.append([arr_[2*n], arr_[2*n+1]])
    visit = [False] * (N+2)
    inner = [[0] * (N+2) for _ in range(0,N+2)]
    MIN = 999999999
    for i in range(0,N+2):
        for j in range(i+1,N+2):
            inner[i][j] = abs(arr[i][0] - arr[j][0]) + abs(arr[i][1] - arr[j][1]) 
            inner[j][i] = abs(arr[i][0] - arr[j][0]) + abs(arr[i][1] - arr[j][1]) 
    
    dfs(0, 0,0)
    print("#" + str(test) + " " + str(MIN))